src/cloudWatchLogs/commands/addLogEvents.ts

```diff
@@ -10,6 +10,8 @@ import { getLogger } from '../../shared/logger/logger'
 
 // TODO: Does this set maxPending for each key to 1, or to all keys?
 // If the latter, we need to make a key registry...
+// TODO: Cut a PR to the async-lock package?...as of now, maxPending = 0 is theoretically ideal, but also falsy:
+// https://github.com/rogierschouten/async-lock/blob/78cb0c2441650d7bdc148548f99542ccc9c93fd7/lib/index.js#L19
 const lock = new AsyncLock({ maxPending: 1 })
```

Comment:
is debounce more appropriate?

Reply:
You might be looking at an old revision? `lock.isBusy` works, but I can also look at debounce. The only thing that doesn't work reliably right now is the busy codelens redraw.

Reply:
Is the goal ...

- to force the event-handing to be serialized (i.e. handle _all_ events, but handle them in-order and one-at-a-time)? 
- to only handle 1 event and discard potentially "redundant" events during some interval? 
  - the "debounce" package/module is useful for this (we already depend on it)
- something else (please describe)?

Reply:
Handle one and discard. The codelens should be clickable one time and that's it (in a perfectly-synchronous world, the fired event would assist this by showing a "Loading..." message).

I'll check debounce.

Reply:
So one thing about debounce: it only works if you're calling the same instance of a function (instead of async-lock, which uses an arbitrary key globally(?)). That gives me a couple options:

* export `addLogEvents` as a debounced function, which means that only one codelens or action (or anything else) can call that function at any given time globally
* create a debounced function in the codelens provider, which means that each codelens will individually be debounced, but any other actions might not be (e.g. a log tail and a codelens click at the same time would produce two API calls).
  * any out-of-order `onDidChangeCodeLensEvent.fire()` calls would cause the debounced functions to be refreshed, potentially dumping them
* add a debounced instance of `addLogEvents` to the registry entries. That would call for a significant retool of the logic here, but would be the most foolproof (since it will be tied to each log individually).

Thoughts? I'm also not 100% sure what testing will look like with this.

Reply:
In favor of whatever requires the fewest moving parts maintained by us, while also meeting the correct semantics.

>     export `addLogEvents` as a debounced function, which means that only one codelens or action (or anything else) can call that function at any given time globally

Or wrap it with a function like:

```
/**
 *  Calls addLogEvents().  Other (asynchronous) calls to this function will be ignored while it is working.
 */
export function addLogEvents_debounced(...) {
  return addLogEvents(...)
}
```

Reply:
I think a lot of my apprehension is coming from implementing log tailing, which I was going to do in a separate PR. Would you mind if I instead add the tailing implementation to this PR, for less potential churn on the debounce logic?

Reply:
No problem :)

Reply:
Will shift tailing logic to a later PR. Internally tracking this discussion and will revisit when adding tailing.